### ISSUE TITLE
Index failing every 5000 blocks exactly

### ISSUE THREAD

**Issue by Bird-NZ** [ordinals/ord#2455]
Still seeing this issue again - https://github.com/ordinals/ord/issues/813:

**Command**:
ord --bitcoin-data-dir=D:\Bitcoin --index-sats index run
**Issue**:
Index failing every 5000 blocks _exactly_.
**Error**:
error: JSON-RPC error: transport error: Couldn't connect to host: An established connection was aborted by the software in your host machine. (os error 10053)
**Ord version**:
0.9.0

**Comment by DrJingLee** [comment]
I have same issue when run
'''
Ord --index-sats index run
'''
Both 0.9.0 and 0.8.3

Previous version run

Ord index run is generally fine.

**Comment by Bird-NZ** [comment]
@casey do you have any advise on this issue?

**Comment by cryptoni9n** [comment]
Subscribing - I see this error in my sleep.  There's a lot of people running into this on the Ordicord, also.

**Comment by cypherpork** [comment]
There is a workaround: if you set `ord` indexing as an OS service, with a `Restart=on-failure`, it can be left unattended, and it will build the index.  Your .service file on Ubuntu will look similar to:

`cat /etc/systemd/system/ord.service`

```
[Unit]
After=bitcoind.service
Description=Ord Server
StartLimitBurst=120
StartLimitIntervalSec=10m

[Service]
AmbientCapabilities=CAP_NET_BIND_SERVICE
Environment=RUST_BACKTRACE=1
Environment=RUST_LOG=info
ExecStart=/usr/local/bin/ord \
  --bitcoin-data-dir /home/bitcoin/.bitcoin/testnet3 \
  --cookie-file /home/bitcoin/.bitcoin/testnet3/.cookie \
  --data-dir /home/bitcoin/bin \
  --config-dir /home/bitcoin/bin \
  --chain test \
  --index-sats \
  server \
  --http --http-port 8000
Group=users
LimitNOFILE=65536
MemoryDenyWriteExecute=true
NoNewPrivileges=true
PrivateDevices=true
PrivateTmp=true
Restart=on-failure
RestartSec=5s
TimeoutStopSec=10m
Type=simple
User=bitcoin
WorkingDirectory=/home/bitcoin/bin/

[Install]
WantedBy=multi-user.target
```
This configuration will do what you need, but I took several security shortcuts, so use it in a sandbox only; beef it up if you plan to use it for a public-facing prod service.

---

The indexing process cycles through two phases: 1) parsing the data it receives from bitcoind via RPC and building the index in the memory and 2) flushing the memory into the database (every 5000 blocks). 
While step 2 is getting executed, RPC waits.
In my experience, the error most frequently happens if phase 2 takes longer than the RPC can wait, i.e., RPC timeouts.

A timeout could happen on the server-side (bitcoind) or the client-side (ord).
On the server side, timeout can be adjusted by adding to your bitcoin.conf `rpcservertimeout=<some ridiculously large number like 1200>`.

I do not know how to adjust RPC timeout on the client side for this setup.

---

A better way to fix it would be to make parsing and flushing parallel processes instead of sequential. Refactoring like this can potentially improve indexing performance. But at the same time, it will increase the code complexity, which makes it not worth the benefit IMHO, considering the existing workarounds.

---

Also, I have noticed that the flushing phase is CPU-intensive (single-core) and produces very little write activity, which urges a desire to look under the hood.  This part could be worth the effort.

**Comment by cryptoni9n** [comment]
Thanks for the information, cyperpork.  I've been doing something similar for Windows in that I've created a looping batch file that will auto-rerun the command once it fails.  My problem is that it takes a very long time to fail - sometimes it just hangs without failing.  What I've been doing is a graceful shutdown once it freezes. I let it sit at the graceful shutdown screen until it times out with the 10053 error, then the batch file rekicks.  When it restarts, it picks up and goes from where it died.  I don't know if this will produce an uncorrupted database or not yet, (day 3 of straight indexing!).

**Comment by cypherpork** [comment]
> I don't know if this will produce an uncorrupted database or not yet, (day 3 of straight indexing!)
@cryptoni9n It seems that when `ord` indexing picks up from where it has left, it checks the integrity of the `index.redb` it has created so far. As long as it is still inching away, it will come out alright.

**Comment by cypherpork** [comment]
In addition to my [earlier_post](https://github.com/ordinals/ord/issues/2455#issuecomment-1752128438), here's the recipe for how to create an ord service on macOS, Ventura+:
```
sudo cat /Library/LaunchDaemons/com.ord.plist

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>EnvironmentVariables</key>
	<dict>
		<key>PATH</key>
		<string>/bin:/usr/bin:/usr/local/bin</string>
		<key>RUST_BACKTRACE</key>
		<string>=1</string>
		<key>RUST_LOG</key>
		<string>=info</string>
	</dict>
	<key>GroupName</key>
	<string>staff</string>
	<key>InitGroups</key>
	<true/>
	<key>KeepAlive</key>
	<dict>
		<key>OtherJobEnabled</key>
		<dict/>
	</dict>
	<key>Label</key>
	<string>com.ordinals.ord</string>
	<key>ProcessType</key>
	<string>Interactive</string>
	<key>Program</key>
	<string>/Users/bitcoin/ordscripts/ord.sh</string>
	<key>RunAtLoad</key>
	<true/>
	<key>StandardErrorPath</key>
	<string>/tmp/ord.log</string>
	<key>StandardOutPath</key>
	<string>/tmp/ord.log</string>
	<key>UserName</key>
	<string>bitcoin</string>
	<key>WorkingDirectory</key>
	<string>/Users/bitcoin/Library/Application Support/ord</string>
</dict>
</plist>

cat /Users/bitcoin/ordscripts/ord.sh

#!/usr/bin/env bash
echo "Starting ord on"
date
echo "waiting for bitcoind to come up ..."
while ! grep -Fqs "init message: Done loading" /tmp/bitcoind.log
do
sleep 1
done
while [ ! -f ~/.bitcoin/.cookie ]
do
echo "waiting for the cookie ..."
  sleep 1 
done
sleep 5
echo "launching ord ..."
# for testnet:
/usr/local/bin/ord --cookie-file ~/.bitcoin/.cookie -t server --http --http-port 8008 --address 0.0.0.0

# To launch:
sudo launchctl bootstrap system /Library/LaunchDaemons/com.ord.plist
# To stop:
sudo launchctl bootout system /Library/LaunchDaemons/com.ord.plist
# To observe:
tail -F /tmp/ord.log
``` 
It comes with the same security disclaimer: *This configuration will do what you need, but I took several security shortcuts, so use it in a sandbox only; beef it up if you plan to use it for a public-facing prod service.*

**Comment by hockma22** [comment]
had the same issue, instead of restrting manually each fail, just put together this batch file below:
run_index.bat
@echo off
:run_command
ord --index-sats index run
if %errorlevel% neq 0 (
    echo Command failed. Retrying...
    timeout /t 5 >nul
    goto run_command
)
echo index succeeded.

**Comment by casey** [comment]
I did some debugging. This error happens after the first commit that takes 90 seconds or more, and the call which triggers the error is in `Reorg::update_savepoints`, which calls `index.client.get_blockchain_info()`. This is the only call that the updater makes to `index.client`. Most RPC calls are made using the fetcher, which uses a hyper client, which doesn't seem to be susceptible to this timeout.

I opened #2650 with a quick fix, which just creates a new client every time. The longer term fix is either to add a hyper backend to the JSON RPC crate, or to try to fix the JSON RPC crate to not timeout.
